aries_vcx/src/protocols/issuance/holder/state_machine.rs

```diff
@@ -172,18 +172,16 @@ impl HolderSM {
             state,
             source_id,
             thread_id,
-        } = self;
+        } = self.clone();
         verify_thread_id(&thread_id, &cim)?;
         let state = match state {
             HolderFullState::Initial(state_data) => match cim {
                 CredentialIssuanceAction::CredentialProposalSend(proposal_data) => {
-                    let proposal = CredentialProposal::from(proposal_data).set_id(&thread_id);
-                    send_message.ok_or(VcxError::from_msg(
+                    let send_message = send_message.ok_or(VcxError::from_msg(
                         VcxErrorKind::InvalidState,
                         "Attempted to call undefined send_message callback",
-                    ))?(proposal.to_a2a_message())
-                    .await?;
-                    HolderFullState::ProposalSent(ProposalSentState::new(proposal))
+                    ))?;
+                    self.send_proposal(proposal_data, send_message).await?.state
```

Comment:
This is quite a code smell, if nothing else changes than state, why only state is not removed? I know the answer, because we use `send_proposal` from `holder.rs` where the entire `SmHolder` is what we need. 
The clean solution I hence advocate here, is to modify the 
```
let state = match state {
```
on line `177` such that the result of the match is entire `HolderSm`, not just state. That way we don't have to do any field cherry picking in situations like on this line. The simpler branches, like `186` will be more verbose but it'll be much cleaner and less error prone.

That being said tho... ideally `handle_message` would not exist in state machine, just like it does not exist for connection state machines.

Reply:
Simplified the function as discussed